File: 2256. Minimum Average Difference/Approach 3 - Prefix Sum Optimized/Minimum Average Difference.cpp
class Solution {
public:
    int minimumAverageDifference(vector<int>& nums) {
        int n = int(nums.size());
        int ans = -1;
        int minAvgDiff = numeric_limits<int>::max();
        long long currPrefixSum = 0;
        
        long long totalSum = 0;
        for (int index = 0; index < n; ++index) {
            totalSum += nums[index];
        }
        
        for (int index = 0; index < n; ++index) {
            currPrefixSum += nums[index];
            
            long long leftPartAverage = currPrefixSum;
            leftPartAverage /= (index + 1);
            
            long long rightPartAverage = totalSum - currPrefixSum;
            if (index != n - 1) {
                rightPartAverage /= (n - index - 1);
            }
            
            int currDifference = int(abs(leftPartAverage - rightPartAverage));
            if (currDifference < minAvgDiff) {
                minAvgDiff = currDifference;
                ans = index;
            }
        }

        return ans;
    }
};

File: Problems & Solutions/2256. Minimum Average Difference/04 December 2022/Approach 3 - Prefix Sum Optimized/Minimum Average Difference.swift
class Solution {
    func minimumAverageDifference(_ nums: [Int]) -> Int {
        let n = nums.count
        var ans = -1
        var minAvgDiff = Int32.max
        var currPrefixSum: Int64 = 0
        
        var totalSum: Int64 = 0
        for index in 0..<n {
            totalSum += Int64(nums[index])
        }
        
        for index in 0..<n {
            currPrefixSum += Int64(nums[index])
            
            var leftPartAverage = currPrefixSum
            leftPartAverage /= Int64(index + 1)
            
            var rightPartAverage = totalSum - currPrefixSum
            if (index != n - 1) {
                rightPartAverage /= Int64(n - index - 1)
            }
            
            let currDifference = abs(leftPartAverage - rightPartAverage)
            if (currDifference < minAvgDiff) {
                minAvgDiff = Int32(currDifference)
                ans = index
            }
        }
        
        return ans;
    }
}